res/controllers/Pioneer-DDJ-SR2-scripts.js

```diff
@@ -0,0 +1,1411 @@
+        }
+    };
+
+    this.gridSetButton = new components.Button({
+        midi: [0x90 + channelOffset, 0x64],
+        key: "beats_translate_curpos"
+    });
+
+    this.keySync = new components.Button({
+        midi: [0x90 + channelOffset, 0x70],
+        key: "sync_key"
+    });
+
+    this.keyShiftDown = new components.Button({
+        midi: [0x90 + channelOffset, 0x72],
+        key: "pitch_down"
+    });
+
+    this.keyShiftUp = new components.Button({
+        midi: [0x90 + channelOffset, 0x73],
+        key: "pitch_up"
+    });
+
+    this.keyReset = new components.Button({
+        midi: [0x90 + channelOffset, 0x71],
+        inKey: "reset_key"
+    });
+    this.wheelRing = new WheelRing({
+        color: DDJSR2.wheelColor[deckNumber],
+        deckNumber: deckNumber,
+        channelOffset: channelOffset
+    });
+
+    // Attach the pads to the deck instance
+    this.padSection = new DDJSR2.PadSection(this, channelOffset);
+};
+
+DDJSR2.Deck.prototype = Object.create(components.Deck.prototype);
+
+
+DDJSR2.MixerContainer = function() {
+    this.channels = [];
+    for (let i = 0; i < 4; i++) {
+        this.channels[i] = new DDJSR2.Channel(i);
+    }
+
+    this.crossfader = new components.Pot({
+        group: "[Master]",
+        inKey: "crossfader",
+    });
+    DDJSR2.panels = [false, false];
+    this.panelSelectButton = function(channel, control, value, _status, _group) {
+        if (value) {
+            if ((DDJSR2.panels[0] === false) && (DDJSR2.panels[1] === false)) {
+                DDJSR2.panels[0] = true;
+            } else if ((DDJSR2.panels[0] === true) && (DDJSR2.panels[1] === false)) {
+                DDJSR2.panels[1] = true;
+            } else if ((DDJSR2.panels[0] === true) && (DDJSR2.panels[1] === true)) {
+                DDJSR2.panels[0] = false;
+            } else if ((DDJSR2.panels[0] === false) && (DDJSR2.panels[1] === true)) {
+                DDJSR2.panels[1] = false;
+            }
+
+            engine.setValue("[Samplers]", "show_samplers", DDJSR2.panels[0]);
+            engine.setValue("[EffectRack1]", "show", DDJSR2.panels[1]);
+        }
+    };
+
+    this.shiftPanelSelectButton = function(channel, control, value, _status, _group) {
+        if (value) {
+            if ((DDJSR2.panels[0] === false) && (DDJSR2.panels[1] === false)) {
+                DDJSR2.panels[1] = true;
+            } else if ((DDJSR2.panels[1] === true) && (DDJSR2.panels[0] === false)) {
+                DDJSR2.panels[0] = true;
+            } else if ((DDJSR2.panels[0] === true) && (DDJSR2.panels[1] === true)) {
+                DDJSR2.panels[1] = false;
+            } else if ((DDJSR2.panels[1] === false) && (DDJSR2.panels[0] === true)) {
+                DDJSR2.panels[0] = false;
+            }
+
+            engine.setValue("[Samplers]", "show_samplers", DDJSR2.panels[0]);
+            engine.setValue("[EffectRack1]", "show", DDJSR2.panels[1]);
+        }
+    };
```

Comment:
unless JS weird semantics about function calling screw me over again, this should work.
```suggestion
    const panelStates = new DDJSR2.RingBufferView([[false, false], [true, false], [true, true], [false, true]]);
    const makePanelSelectButton = method => function(channel, control, value, _status, _group) { 
            if (value) {
                const [showSamplers, showFX] = method();
    
                engine.setValue("[Samplers]", "show_samplers", showSamplers);
                engine.setValue("[EffectRack1]", "show", showFX);
            }
        };
    this.panelSelectButton = makePanelSelectButton(panelStates.next);
    this.shiftPanelSelectButton = makePanelSelectButton(panelStates.previous);
```